src/compiler/checker.ts

```diff
@@ -3328,25 +3329,62 @@ namespace ts {
             }
         }
 
-        // Returns true if the interface given by the symbol is free of "this" references. Specifically, the result is
-        // true if the interface itself contains no references to "this" in its body, if all base types are interfaces,
-        // and if none of the base interfaces have a "this" type.
-        function isIndependentInterface(symbol: Symbol): boolean {
+        function isIndependentBaseInterfaceList(nodes: NodeArray<ExpressionWithTypeArguments>): boolean {
+            if (!nodes) {
+                return true;
+            }
+            for (const node of nodes) {
+                if (!isIndependentHeritageClauseElement(node, /*classSymbol*/ undefined)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        function isIndependentHeritageClauseElement(node: ExpressionWithTypeArguments, classSymbol: Symbol): boolean {
+            if (!node) {
+                return true;
+            }
+            if (isSupportedExpressionWithTypeArguments(node)) {
+                const baseSymbol = resolveEntityName(node.expression, SymbolFlags.Type, /*ignoreErrors*/ true);
+                if (!baseSymbol || baseSymbol === unknownSymbol) {
+                    // pessimistically asssume that base type is not independent
+                    return false;
+                }
+                const baseType = <InterfaceType>getDeclaredTypeOfSymbol(baseSymbol);
+                if (baseType.thisType) {
+                    return false;
+                }
+            }
+            else if (classSymbol) {
+                Debug.assert((classSymbol.flags & SymbolFlags.Class) !== 0);
+                const baseTypes = getBaseTypes(<InterfaceType>getDeclaredTypeOfClassOrInterface(classSymbol));
+                if (baseTypes.length) {
+                    return (<InterfaceType>baseTypes[0]).thisType === undefined;
+                }
+            }
+            return true;
+        }
+
+        // Returns true if the class\interface given by the symbol is free of "this" references. Specifically, the result is
+        // true if the class\interface itself contains no references to "this" in its body, if all base types are classes\interfaces,
+        // and if none of the base classes\interfaces have a "this" type.

```

Comment:
For interfaces, what if the derived types (if any) have references to `this`? Wouldn't the base interface need a `this` type then?